File: bench/immutable/immutable.go
// immutable.go — a particle step in the immutable-update idiom: every step
// replaces each record wholesale instead of mutating fields. The canonical
// functional-state kernel (reducers, persistent game state, event-sourced
// models). Go gets the pattern free through value semantics — a struct
// assigned by value, zero allocation — the static reference for what the
// fresh-object JS idiom costs. Pure integer bounce physics, so positions are
// bit-identical across every engine and native target.
//
// Reports: median ms across N_RUNS, throughput in particle-steps/µs, FNV-1a
// checksum over the per-pass position folds.
package main

import (
	"fmt"
	"time"
)

const (
	n       = 4096 // particles
	steps   = 32   // steps per pass (one record replacement per particle per step)
	lim     = 1023 // box bound
	nRuns   = 21
	nWarmup = 5
)

type p struct{ x, y, vx, vy int32 }

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

// Deterministic initial state — XorShift32 positions in [0, LIM], velocities
// in [-8, 8] forced nonzero, identical per target.
func initParticles(ps []p) {
	s := int32(0x1234abcd)
	for i := 0; i < n; i++ {
		s ^= s << 13
		s ^= int32(uint32(s) >> 17)
		s ^= s << 5
		vx := int32((uint32(s)>>4)&15) - 8
		vy := int32((uint32(s)>>8)&15) - 8
		if vx == 0 {
			vx = 1
		}
		if vy == 0 {
			vy = 1
		}
		ps[i] = p{int32((uint32(s) >> 12) & lim), int32((uint32(s) >> 20) & lim), vx, vy}
	}
}

func runKernel(ps []p) uint32 {
	h := uint32(0x811c9dc5)
	for it := 0; it < steps; it++ {
		sum := int32(it)
		for i := 0; i < n; i++ {
			q := ps[i]
			nx, ny := q.x+q.vx, q.y+q.vy
			hitX := nx < 0 || nx > lim
			hitY := ny < 0 || ny > lim
			x, y := nx, ny
			vx, vy := q.vx, q.vy
			if hitX {
				x, vx = q.x, -q.vx
			}
			if hitY {
				y, vy = q.y, -q.vy
			}
			ps[i] = p{x, y, vx, vy}
			sum = int32(uint32(sum) + uint32(x) + uint32(y)*31)
		}
		h = mix(h, uint32(sum))
	}
	return h
}

func main() {
	ps := make([]p, n)
	var cs uint32
	// Fresh particle set per run — the kernel mutates the array, so timing runs
	// must not compound each other's motion.
	for i := 0; i < nWarmup; i++ {
		initParticles(ps)
		cs = runKernel(ps)
	}

	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		initParticles(ps)
		t0 := time.Now()
		cs = runKernel(ps)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), cs, n*steps, 1, nRuns)
}

File: bench/immutable/immutable.rs
// immutable.rs — a particle step in the immutable-update idiom: every step
// replaces each record wholesale instead of mutating fields. The canonical
// functional-state kernel (reducers, persistent game state, event-sourced
// models). Rust gets the pattern free through value semantics — a Copy struct
// assigned by value, zero allocation — the static reference for what the
// fresh-object JS idiom costs. Pure integer bounce physics, so positions are
// bit-identical across every engine and native target.
//
// Reports: median ms across N_RUNS, throughput in particle-steps/µs, FNV-1a
// checksum over the per-pass position folds.
use std::time::Instant;

const N: usize = 4096;           // particles
const STEPS: usize = 32;         // steps per pass (one record replacement per particle per step)
const LIM: i32 = 1023;           // box bound
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

#[derive(Clone, Copy)]
struct P {
    x: i32,
    y: i32,
    vx: i32,
    vy: i32,
}

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

// Deterministic initial state — XorShift32 positions in [0, LIM], velocities
// in [-8, 8] forced nonzero, identical per target.
fn init_particles() -> Vec<P> {
    let mut ps = Vec::with_capacity(N);
    let mut s: i32 = 0x1234abcd_u32 as i32;
    for _ in 0..N {
        s ^= s << 13;
        s ^= ((s as u32) >> 17) as i32; // >>> 17 (unsigned right shift)
        s ^= s << 5;
        let vx = (((s as u32) >> 4) & 15) as i32 - 8;
        let vy = (((s as u32) >> 8) & 15) as i32 - 8;
        ps.push(P {
            x: (((s as u32) >> 12) & LIM as u32) as i32,
            y: (((s as u32) >> 20) & LIM as u32) as i32,
            vx: if vx == 0 { 1 } else { vx },
            vy: if vy == 0 { 1 } else { vy },
        });
    }
    ps
}

fn run_kernel(ps: &mut [P]) -> u32 {
    let mut h: u32 = 0x811c9dc5;
    for it in 0..STEPS {
        let mut sum: i32 = it as i32;
        for i in 0..N {
            let p = ps[i];
            let nx = p.x + p.vx;
            let ny = p.y + p.vy;
            let hit_x = nx < 0 || nx > LIM;
            let hit_y = ny < 0 || ny > LIM;
            let x = if hit_x { p.x } else { nx };
            let y = if hit_y { p.y } else { ny };
            let vx = if hit_x { -p.vx } else { p.vx };
            let vy = if hit_y { -p.vy } else { p.vy };
            ps[i] = P { x, y, vx, vy };
            sum = sum.wrapping_add(x).wrapping_add(y.wrapping_mul(31));
        }
        h = mix(h, sum as u32);
    }
    h
}

fn main() {
    let mut cs = 0u32;
    // Fresh particle set per run — the kernel mutates the array, so timing runs
    // must not compound each other's motion.
    for _ in 0..N_WARMUP {
        let mut ps = init_particles();
        cs = run_kernel(&mut ps);
    }

    let mut samples = [0.0f64; N_RUNS];
    for s in &mut samples {
        let mut ps = init_particles();
        let t0 = Instant::now();
        cs = run_kernel(&mut ps);
        *s = t0.elapsed().as_secs_f64() * 1000.0;
    }

    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples),
        cs,
        N * STEPS,
        1,
        N_RUNS
    );
}
